py/dml/c_backend.py

```diff
@@ -2258,10 +2293,110 @@ def generate_alloc_trait_vtables(node):
         if subnode.traits:
             for trait in subnode.traits.referenced:
                 name = subnode.traits.vtable_cname(trait)
-                out(f'{name} = MM_MALLOC(1, typeof(*{name}));\n')
+                out(f'{name} = MM_ZALLOC(1, typeof(*{name}));\n')
     out('}\n', preindent=-1)
     splitting_point()
 
+def generate_startup_call_loops(startup_methods):
+    by_dims = {}
+    for (method_data_tuple, dims) in startup_methods:
+        by_dims.setdefault(dims, []).append(method_data_tuple)
+    for dims in by_dims:
+        for i in range(len(dims)):
+            idxvar = '_i%d' % (i,)
+            out('for (uint32 %s = 0; %s < %d; %s++) {\n'
+                % (idxvar, idxvar, dims[i], idxvar), postindent=1)
+
+        for (site, gen_call, data) in by_dims[dims]:
+            idxvars = ['_i%d' % (i,) for i in range(len(dims))]
+            gen_call(data, idxvars)
+
+        for i in range(len(dims)):
+            out('}\n', preindent=-1)
+
+def generate_startup_calls_entry_function(devnode):
+    start_function_definition('void _startup_calls(void)')
+    out('{\n', postindent=1)
+
+    def generate_regular_call(method, idxvars):
```

Comment:
you could move out this one (and `generate_trait_calls`) to global scope as well